x/ccv/provider/keeper/partial_set_security.go

```diff
@@ -0,0 +1,148 @@
+		}
+
+		err = k.AssignConsumerKey(ctx, chainID, validator, consumerTMPublicKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// HandleOptOut prepares validator `providerAddr` to opt out from running `chainID`.
+// Note that the validator only opts out at the end of an epoch.
+func (k Keeper) HandleOptOut(ctx sdk.Context, chainID string, providerAddr types.ProviderConsAddress) error {
+	if _, found := k.GetConsumerClientId(ctx, chainID); !found {
+		// A validator can only opt out from a running chain. We check this by checking the consumer client id, because
+		// `SetConsumerClientId` is set when the chain starts in `CreateConsumerClientInCachedCtx` of `BeginBlockInit`.
+		return errorsmod.Wrapf(
+			types.ErrUnknownConsumerChainId,
+			"opting out of an unknown or not running consumer chain, with id: %s", chainID)
+	}
+
+	if topN, found := k.GetTopN(ctx, chainID); found && topN > 0 {
+		// a validator cannot opt out from a Top N chain if the validator is in the Top N validators
+		validator, validatorFound := k.stakingKeeper.GetValidatorByConsAddr(ctx, providerAddr.ToSdkConsAddr())
+		if !validatorFound {
+			return errorsmod.Wrapf(
+				stakingtypes.ErrNoValidatorFound,
+				"validator with consensus address %s could not be found", providerAddr.ToSdkConsAddr())
+		}
+		power := k.stakingKeeper.GetLastValidatorPower(ctx, validator.GetOperator())
+		minPowerToOptIn := k.ComputeMinPowerToOptIn(ctx, chainID, k.stakingKeeper.GetLastValidators(ctx), topN)
+
+		if power >= minPowerToOptIn {
+			return errorsmod.Wrapf(
+				types.ErrCannotOptOutFromTopN,
+				"validator with power (%d) cannot opt out from Top N chain because all validators"+
+					"with at least %d power have to validate", power, minPowerToOptIn)
+		}
+	}
+
+	k.DeleteOptedIn(ctx, chainID, providerAddr)
+	return nil
+}
+
+// OptInTopNValidators opts in to `chainID` all the `bondedValidators` that have at least `minPowerToOptIn` power
+func (k Keeper) OptInTopNValidators(ctx sdk.Context, chainID string, bondedValidators []stakingtypes.Validator, minPowerToOptIn int64) {
+	for _, val := range bondedValidators {
+		power := k.stakingKeeper.GetLastValidatorPower(ctx, val.GetOperator())
+		if power >= minPowerToOptIn {
+			consAddr, err := val.GetConsAddr()
+			if err != nil {
+				k.Logger(ctx).Error("could not retrieve validators consensus address",
+					"validator", val,
+					"error", err)
+				continue
+			}
+
+			// if validator already exists it gets overwritten
+			k.SetOptedIn(ctx, chainID, types.NewProviderConsAddress(consAddr))
+		} // else validators that do not belong to the top N validators but were opted in, remain opted in
+	}
+}
+
+// ComputeMinPowerToOptIn returns the minimum power needed for a validator (from the bonded validators)
+// to belong to the `topN` validators. `chainID` is only used for logging purposes.
+func (k Keeper) ComputeMinPowerToOptIn(ctx sdk.Context, chainID string, bondedValidators []stakingtypes.Validator, topN uint32) int64 {
```

Comment:
Good idea. I'll work on that

Reply:
There is a PR for it here: https://github.com/cosmos/interchain-security/pull/1840
But in retrospect I think it's not necessary for now, and I'm not confident making the change now, since it touches on pretty core logic.
It's something that could be added later if we deem it necessary; it needs a migration either way.